OpenCL/m24000-pure.cl

```diff
@@ -0,0 +1,881 @@
+
+  u32 iv[4] = { 0 };
+
+  u32 res[20]; // full would be 24 x u32 (96 bytes)
+
+  if (version == '8')
+  {
+
+    #define KEYLEN 60
+
+    u32 ks[KEYLEN];
+
+    AES256_set_decrypt_key (ks, ctx.opad.h, s_te0, s_te1, s_te2, s_te3, s_td0, s_td1, s_td2, s_td3);
+
+    for (u32 i = 0; i < 20; i += 4) // 96 bytes output would contain the full 32 byte checksum
+    {
+      u32 data[4];
+
+      data[0] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 0];
+      data[1] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 1];
+      data[2] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 2];
+      data[3] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 3];
+
+      u32 out[4];
+
+      aes256_decrypt (ks, data, out, s_td0, s_td1, s_td2, s_td3, s_td4);
+
+      res[i + 0] = hc_swap32_S (out[0] ^ iv[0]);
+      res[i + 1] = hc_swap32_S (out[1] ^ iv[1]);
+      res[i + 2] = hc_swap32_S (out[2] ^ iv[2]);
+      res[i + 3] = hc_swap32_S (out[3] ^ iv[3]);
+
+      iv[0] = data[0];
+      iv[1] = data[1];
+      iv[2] = data[2];
+      iv[3] = data[3];
+    }
+  }
+
+
+  if (version == '9')
+  {
+    u32 sk[4];
+    u32 lk[40];
+
+    ctx.opad.h[0] = hc_swap32_S (ctx.opad.h[0]);
+    ctx.opad.h[1] = hc_swap32_S (ctx.opad.h[1]);
+    ctx.opad.h[2] = hc_swap32_S (ctx.opad.h[2]);
+    ctx.opad.h[3] = hc_swap32_S (ctx.opad.h[3]);
+    ctx.opad.h[4] = hc_swap32_S (ctx.opad.h[4]);
+    ctx.opad.h[5] = hc_swap32_S (ctx.opad.h[5]);
+    ctx.opad.h[6] = hc_swap32_S (ctx.opad.h[6]);
+    ctx.opad.h[7] = hc_swap32_S (ctx.opad.h[7]);
+
+    twofish256_set_key (sk, lk, ctx.opad.h);
+
+    for (u32 i = 0; i < 20; i += 4) // 96 bytes output would contain the full 32 byte checksum
+    {
+      u32 data[4];
+
+      data[0] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 0];
+      data[1] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 1];
+      data[2] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 2];
+      data[3] = esalt_bufs[DIGESTS_OFFSET].ciphertext[i + 3];
+
+
+      u32 out[4];
+
+      twofish256_decrypt (sk, lk, data, out);
+
+      res[i + 0] = hc_swap32_S (out[0] ^ iv[0]);
+      res[i + 1] = hc_swap32_S (out[1] ^ iv[1]);
+      res[i + 2] = hc_swap32_S (out[2] ^ iv[2]);
+      res[i + 3] = hc_swap32_S (out[3] ^ iv[3]);
+
+      iv[0] = data[0];
+      iv[1] = data[1];
+      iv[2] = data[2];
+      iv[3] = data[3];
+    }
+  }
+
+  if (version == 'a')
+  {
+    printf("%s\n", "SERPENT");
+    u32 ks_serpent[140];
+    ctx.opad.h[0] = hc_swap32_S (ctx.opad.h[0]);
+    ctx.opad.h[1] = hc_swap32_S (ctx.opad.h[1]);
+    ctx.opad.h[2] = hc_swap32_S (ctx.opad.h[2]);
+    ctx.opad.h[3] = hc_swap32_S (ctx.opad.h[3]);
+    ctx.opad.h[4] = hc_swap32_S (ctx.opad.h[4]);
+    ctx.opad.h[5] = hc_swap32_S (ctx.opad.h[5]);
+    ctx.opad.h[6] = hc_swap32_S (ctx.opad.h[6]);
+    ctx.opad.h[7] = hc_swap32_S (ctx.opad.h[7]);
+    serpent256_set_key (ks_serpent, ctx.opad.h);
+    for (u32 i = 0; i < 20; i += 4) // 96 bytes output would contain the full 32 byte checksum
```

Comment:
code style/formatting: we normally add some newlines here between the blocks (after `u32 ks_serpent[140];` before **and** after `serpent256_set_key (ks_serpent, ctx.opad.h);`